### ISSUE TITLE
IMap Source for Jet SQL

### ISSUE THREAD

**Issue by robinroos** [hazelcast/hazelcast#18539]
This specifically relates to Hazelcast JET, and the SQL implementation which it implements as pipelines.

In order to implement an efficient and stable query engine over data stored in IMaps I request that Jet SQL be able to 

`SELECT ... FROM MAP "mapName" ... `

with mapName being the name of an IMap.

The usual SQL clauses should be supported, including but not limited to WHERE, GROUP BY, projection, aggregations etc.

I am happy to adopt a specific serialization style (e.g. Portable) if that is helpful at runtime.  Presumably the SQL implementation of FROM MAP would be agnostic to that and the efficiencies would accrue lower-down in the stack where predicates are evaluated.

My specific use-case is as follows.  

I am re-implementing a generic Query Service over the data stored in IMaps.  The Query Service will receive JSON documents which describe the query and also contain a reply-to JMS Temporary Destination and a textual correlationId (which ties the response back to the request in the client space).  In order to be stable, scalable and performant the Query Service must be stateless.  The design intention is to parse the JSON query definition into a Hazelcast JET SQL statement and submit that statement for execution.  Queries may be "Batch" or "Streaming".  Query execution must be asynchronous in order that no state needs to be maintained by the Query Service.  Once the query is submitted it becomes the responsibility of Hazelcast JET to evaluate the query (using FROM MAP) and stream results to the JMS Destination (using WRITE TO JMS).

Query Service is also dependent on a separate feature request, "WRITE TO JMS - JMS Sink for Jet SQL", which will publish the projected columns as a JSON map.

The correlationId will be provided into the Hazelcast JET SQL statement using `AS` and will appear verbatim in the projection.  By way of example:
 
Query:
`SELECT "12345" AS correlatiionId, columnName1, columnName2 ..., FROM MAP "mapName" ... WHERE ... GROUP BY ... WRITE TO JMS("destinationName")`

JMS text message payload (one message per row)
```
{
    "correlationId" : "12345"
    "columnName1" : "columnValue1", 
    "columnName2" : "columnValue2",
    ...
}
```

This service will be a very powerful example of how Hazelcast JET SQL can be used, but it cannot be implemented (at least not in-line with the stated Design Intention) without IMap as a Source for Hazelcast JET SQL.

**Comment by robinroos** [comment]
### Further Considerations: KEY

Each IMap entry comprises key and value.  The fields and expressions expressed in the SELECT clause are presumably taken from the Value (object).  It would be extremely useful to be able to expose also the Key (object).

That would require a token such as "__key" to be expressed in the SELECT clause, e.g.:

`SELECT __key FROM MAP mapName`

`SELECT __key, column1, column2 FROM MAP mapName`

Of course, things can then get more interesting:

`SELECT MAX(__key) FROM MAP mapName`

`SELECT __key.city, __key.country FROM MAP mapName`

Although those particular two are not a specific use-cases of mine at this time!

**Comment by robinroos** [comment]
**Further Considerations: ACTION**

One can conceive of a SELECT ... FROM MAP as providing the client with sufficient information to re-create that subset of the IMap in its own heap space.

A Batch query, which examines a snapshot of the state of the IMap and which ends after processing all (matching) rows always and only represents the Insert of data items into that notional Map in the client's heap space.

A Streaming (Continuous) query, however, reflects changes occurring to the map, the underlying mechanism being either event listeners or the IMap journal.  This introduces the requirement to reflect Removal alongside Upsert (Insert / Update) of items.

The SELECT syntax FROM MAP should thus be able to include a column, perhaps named "__action", which reflects a constant from the set { INSERT , UPDATE, DELETE } or from the set { UPSERT , DELETE }.

Only with access to __action can a Client use the results of a Streaming SELECT ... FROM MAP in order to maintain an accurate subset of the IMap in its heap space.

**Comment by Fly-Style** [comment]
Closed by #18865
